server/notebooks/tasks.py

```diff
@@ -0,0 +1,50 @@
+from collections import defaultdict
+from datetime import datetime, timedelta
+
+import pytz
+from spinach import Tasks
+
+from ..settings import NOTEBOOK_REVISION_FIXED_WINDOW_LENGTH_SECS
+from .models import NotebookRevision
+
+tasks = Tasks()
+
+
+@tasks.task(name="notebooks:execute_notebook_revisions_cleanup")
+def execute_notebook_revisions_cleanup(notebook_id, now_utc=None):
+    """Prue revision hisotry.
+
+    Prune revision history so that there is only one revision in each
+    time window, also marking draft revisions that older than window length
+    as non-draft.
+
+    * Time window: This task groups draft revisions into fixed-size windows
+      (alsocalled Tumbling windows).
+    """
+    draft_revisions = NotebookRevision.objects.filter(notebook_id=notebook_id, is_draft=True)
+    now_utc = now_utc or datetime.now(tz=pytz.utc)
+
+    # group notebook revisions into fixed-size windows
+    revision_groups = defaultdict(list)
+    for revision in draft_revisions:
+        key = int(revision.created.timestamp()) // NOTEBOOK_REVISION_FIXED_WINDOW_LENGTH_SECS
+        revision_groups[key].append(revision)
+    revision_windows = [revision_groups[k] for k in sorted(revision_groups.keys(), reverse=True)]
+
+    threshold = timedelta(seconds=NOTEBOOK_REVISION_FIXED_WINDOW_LENGTH_SECS)
+    # delete the intermediate revisions
+    intermediate_revision_ids = [
+        revision.id
+        for revision_window in revision_windows
+        for revision in revision_window[1:]
+        if now_utc - revision.created >= threshold
+    ]
+    NotebookRevision.objects.filter(id__in=intermediate_revision_ids).delete()
```

Comment:
To be consistent with my suggestion below, I'd suggest doing the filtering on time in this statement, rather than above. So something like:

```py
intermediate_revision_ids = [
    revision.id
    for revision_window in revision_windows
    for revision in revision_window[1:]
]
NotebookRevision.objects.filter(id__in=intermediate_revision_ids, created__lt=now_utc - threshold).delete()
```

Reply:
Please correct me if I am wrong here. I understand that we want to [do database work in the database rather than in Python](https://docs.djangoproject.com/en/2.2/topics/db/optimization/#do-database-work-in-the-database-rather-than-in-python). If I am reading data from database, I prefer to put filters to database layer (filter pushdown), so that I will get less data.

But in above case, we already have the data in memory. so I did the filter in Python and used `NotebookRevision.objects.filter(id__in=small_set_of_ids).delete()` instead of `NotebookRevision.objects.filter(id__in=big_set_of_ids, created__lt=now_utc - threshold).delete()`. The assumption here is database is harder to scale than stateless application.

I am fine with both, I don't have strong preference here.

Reply:
@zzl0 it really depends on what's going on and where the bottlenecks are. In general performing an operation at the database level is going to be an order of magnitude more efficient than doing so in python code -- though in this case I'm suggesting the change mainly for readability reasons (it's easier to see what's going on in a filter expression than a list comprehension)